### ISSUE TITLE
Incoherent validation results

### ISSUE THREAD

**Issue by derreter** [sissaschool/xmlschema#164]
As can be analyzed in the test results, there are some problems with the validation in the different modes:
  - In "iter_errors" mode, with lazy = True, does not validate Constraints.
  - The validation of the decimal that I explain, without the line number, it is impossible to find the wrong value. (Only with e.reason, of course)
  - The other types of validation, do not validate Constraints

Testes and results:

```pycon
>>> for e in xmlschema.iter_errors("XPTO.XML", schema=xs): e.reason
"[<class 'decimal.ConversionSyntax'>]"
XMLSchemaValueError("Value ('01095',) not found for Xsd11Unique(name='ns:CustomerIDConstraint') (2 times)")
```

```pycon
>>> st = dt.now(); len(list(xmlschema.iter_errors("XPTO.XML", schema=xs))), dt.now() - st
(2, datetime.timedelta(seconds=12, microseconds=311921))
>>> st = dt.now(); len(list(xmlschema.iter_errors("XPTO.XML", schema=xs, lazy=True))), dt.now() - st
(1, datetime.timedelta(seconds=7, microseconds=873477))
```

```pycon
>>> xmlschema.validate("XPTO.XML", schema=xs, lazy=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python38\lib\site-packages\xmlschema\documents.py", line 100, in validate
    schema.validate(source, path, schema_path, use_defaults, namespaces)
  File "C:\Python38\lib\site-packages\xmlschema\validators\schema.py", line 1249, in validate
    raise error
xmlschema.validators.exceptions.XMLSchemaDecodeError: failed validating '437,17' with XsdAtomicBuiltin(name='xs:decimal'):

Reason: [<class 'decimal.ConversionSyntax'>]

Schema:

  <xs:simpleType xmlns:xs="http://www.w3.org/2001/XMLSchema" name="{http://www.w3.org/2001/XMLSchema}decimal" id="{http://www.w3.org/2001/XMLSchema}decimal" />

Path: /AuditFile/MasterFiles
```

```pycon
>>> xmlschema.validate("XPTO.XML", schema=xs, path="*")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python38\lib\site-packages\xmlschema\documents.py", line 100, in validate
    schema.validate(source, path, schema_path, use_defaults, namespaces)
  File "C:\Python38\lib\site-packages\xmlschema\validators\schema.py", line 1249, in validate
    raise error
xmlschema.validators.exceptions.XMLSchemaDecodeError: failed validating '437,17' with XsdAtomicBuiltin(name='xs:decimal'):

Reason: [<class 'decimal.ConversionSyntax'>]

Schema:

  <xs:simpleType xmlns:xs="http://www.w3.org/2001/XMLSchema" name="{http://www.w3.org/2001/XMLSchema}decimal" id="{http://www.w3.org/2001/XMLSchema}decimal" />

Instance:

  <ClosingDebitBalance xmlns="urn:OECD:StandardAuditFile-Tax:PT_1.04_01">437,17</ClosingDebitBalance>

Path: /AuditFile/MasterFiles/GeneralLedgerAccounts/Account/ClosingDebitBalance
```

thanks.

**Comment by brunato** [comment]
Hi,

about *xs:decimal* decoding error, the value should be saved to ``e.obj`` attribute. 
The `` [<class 'decimal.ConversionSyntax'>]`` reason is an error due to an unexpected situation that has to be fixed.

Information about line is not available with the *ElementTree* library but the part can be retrievable using the XPath. 

About 'lax' validation and constraints with *iter_errors*: I will add tests to find where the check is missing. The key references are checked at the end of the validation, but with lazy validation this check is done in two steps and has to be done after the 2nd validation step. Probably the missing value is collected but not checked at the end.
Thanks

**Comment by derreter** [comment]
Thank you for your attention,

I am looking for, is a validator capable of validating all the rules of the schema, but also facilitating the verification of the error in the exact place where it occurred. 
Hence I have done the tests that I presented. 
The mode that comes closest to what I need is "iter_errors". If this method validates all the rules of the schema and if it is possible to include the line number in some type of error, it was perfect.

Thanks.

**Comment by brunato** [comment]
For having source line info you can try to pass a lxml's tree to iter_errors, eg.:

```python
import lxml.etree as etree
root = etree.parse("XPTO.XML").getroot()
errors = [e for e in schema.iter_errors(root)]
```
Generated errors contain info about source line (use the property *e.sourceline*).

**Comment by brunato** [comment]
Hi @derreter,

the new release will contains a fix for this. 
Currently i added tests for lazy level 1 (lazy root element), comparing *iter_errors()* results for full and for lazy validation. 

For making lazy validation works for more level some improvements are needed on XPath (*elementpath* library) because in some cases the search path of an element (eg. in case of an element substitute) does not match the right XSD element or XSD type.

**Comment by brunato** [comment]
Patch for simple lazy validation (lazy level 1) is available with release v1.1.1+.
